File: src/omnia/data/structure/mutable/MutableMap.java
package omnia.data.structure.mutable;

import java.util.Optional;
import java.util.function.Supplier;
import omnia.data.structure.Map;

public interface MutableMap<K, V> extends Map<K, V> {

  void putMapping(K key, V value);

  V putMappingIfAbsent(K key, Supplier<V> value);

  default Optional<V> removeKey(K key) {
    return removeUnknownTypedKey(key);
  }

  Optional<V> removeUnknownTypedKey(Object key);

  @Override MutableSet<Entry<K, V>> entries();

  static <K, V> MutableMap<K, V> masking(java.util.Map<K, V> javaMap) {
    return new MaskingMap<>(javaMap);
  }
}

File: src/omnia/data/structure/mutable/MutableMap.kt
package omnia.data.structure.mutable

import omnia.data.structure.Map
import omnia.data.structure.Set
import java.util.Optional
import java.util.function.Supplier

interface MutableMap<K, V> : Map<K, V> {
    fun putMapping(key: K, value: V)
    fun putMappingIfAbsent(key: K, value: Supplier<V>): V
    fun removeKey(key: K): Optional<V> {
        return removeUnknownTypedKey(key)
    }

    fun removeUnknownTypedKey(key: Any?): Optional<V>

    override fun entries(): Set<Map.Entry<K, V>>

    companion object {
        fun <K, V> masking(javaMap: kotlin.collections.MutableMap<K, V>): MutableMap<K, V> {
            return MaskingMap(javaMap)
        }
    }
}